### ISSUE TITLE
How to Parse Yolo Segmentation Output

### ISSUE THREAD

**Issue by Habib-Rehmn** [ultralytics/ultralytics#9020]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am using a yolo segmentation model v8s. My model has only 1 class(sky). Here is the output shape of my model.
I have converted this model to .tflite using yolo.

name: Identity_1
tensor: float32[1,160,160,32]
Mask protos
location: 501

it appears to be an image of 160 by 160 with 32 different values at each pixel

here is some portion of the output

[[[[-0.07753567397594452, -0.20710693299770355, 0.23461437225341797, 0.5300124287605286, 0.24696356058120728, 0.20757406949996948,

how can i parse this output to make a final image to show where the sky is detected.

### Additional

![best_float16 tflite](https://github.com/ultralytics/ultralytics/assets/124852616/18203bfa-ba76-4290-ad79-802389749797)

**Comment by glenn-jocher** [comment]
@Habib-Rehmn hey there! It seems like you're working on a fascinating project with YOLOv8 for image segmentation. From what you've described, it looks like you have a model output with a shape of `[1, 160, 160, 32]`, where each pixel has 32 values associated with it, likely representing the features learned by the network.

To visualize your segmentation, particularly for the sky class, you'd transform these features into a binary mask denoting sky presence. Here's a basic outline to get you started:

1. **Apply a function to convert your model's output to a binary mask**: This typically involves picking a channel if your model predicts a channel per class or applying a threshold if your output is a probability map.

   Since you mentioned having only one class (sky) and provided output shape `[1,160,160,32]`, it's likely you'll need to aggregate these 32 values to decide each pixel's class. Without knowing the exact architecture or method, a common approach might be to apply a softmax operation followed by thresholding to obtain a binary map.

2. **Visualize the binary mask**: Once you have a binary mask, you can use libraries like `matplotlib` or `cv2` to overlay this mask on your original image or simply display the mask.

Here's an example assuming your output is a probability map:

```python
import numpy as np
import matplotlib.pyplot as plt

# Assuming `output` is your model output with shape (1, 160, 160, 32)

# Step 1: Convert to binary mask. This is a simplistic example to illustrate the process.
# You might need a softmax followed by aggregation dependent on your architecture.
binary_mask = np.mean(output, axis=-1) > 0.5  # A placeholder operation

# Step 2: Visualize the mask
plt.imshow(binary_mask[0], cmap='gray')  # Assuming the first axis is batch size
plt.show()
```

Keep in mind, the exact processing depends on how your model predicts the 'sky' class, so you may need to adjust the thresholding step accordingly. If you have further info on how your model's output corresponds to the class prediction, we could provide more specific guidance. Happy coding! 😊

**Comment by Habib-Rehmn** [comment]
import numpy as np
from PIL import Image

# Sample output from the model
output_2 = np.array([[[[0.9476574063301086, 0.6402849555015564, 0.04089251160621643, -0.040269576013088226...]]]])


# Sigmoid function
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Range of threshold values
threshold = 0.50


# Reshape the output to (160, 160, 32) for easier manipulation
output_reshaped = output_2.reshape((160, 160, 32))

# Apply sigmoid function to the output
output_sigmoid = sigmoid(output_reshaped)

# average the sigmoid outputs
average_mask = np.mean(output_sigmoid, axis=-1)

# Then apply thresholding
final_mask = (average_mask > threshold).astype(np.uint8)

# Convert to PIL image
binary_image = Image.fromarray(final_mask * 255)

# Save the binary image
save_path = "./pic3/final_mask.png"
binary_image.save(save_path)

# Inform the user that the final image has been saved
print("Final image saved successfully at:", save_path)    


here is the obtained mask and the ground truth 
 
![final_mask](https://github.com/ultralytics/ultralytics/assets/124852616/25330993-94d0-4d31-bc77-759b53571e70)
![image0](https://github.com/ultralytics/ultralytics/assets/124852616/4c1ad58f-d173-4493-9c83-b1553e57baed)

**Comment by glenn-jocher** [comment]
@Habib-Rehmn hey there! 🌤️ It looks like you're well on your way with segmenting the sky in your images. Your approach of using a sigmoid function to scale the model's output, followed by averaging and thresholding to create a binary mask, seems solid. Here's a slightly streamlined version of your code:

```python
import numpy as np
from PIL import Image

# Assuming output_2 is your model's output
output_2 = np.random.randn(1, 160, 160, 32)  # Random output for demonstration

# Simplify the operations
sigmoid_output = 1 / (1 + np.exp(-output_2)).mean(axis=-1)
mask = (sigmoid_output.squeeze() > 0.5).astype(np.uint8) * 255

# Save as an image
Image.fromarray(mask).save("./final_mask.png")
print("Final image saved successfully!")
```

Your obtained versus ground truth visual comparison should give you good insight into how well your model is doing. Keep tweaking your threshold value to improve fidelity. Great work and keep it up! 🚀

**Comment by Habib-Rehmn** [comment]
Thanks @glenn-jocher for your help. I Implemented the same logic as you do but still not getting correct mask. Also tried many threshold values.

**Comment by Habib-Rehmn** [comment]
I wonder how yolo handle this thing?

**Comment by glenn-jocher** [comment]
@Habib-Rehmn hey there! 🌞 It looks like your YOLOv8 model is providing two types of outputs - one for object detection (coordinates, class labels, and confidence scores) and another for segmentation (mask protos). Yes, both outputs are indeed utilized for different purposes:

1. **Identity (Detection Output):** Use this output for identifying the location (`[x, y, width, height]`), class, and confidence level of detected objects within the image. This is typically used for bounding box predictions.

2. **Identity_1 (Segmentation Output):** This output provides pixel-wise segmentation masks indicating the class presence probability across the image. You would process this output differently from detection output to generate segmentation maps.

For processing the segmentation output to a final image showing where the sky is detected, you've got the right idea: normalize this output to get probabilities, apply a threshold to make a decision for each pixel, and create a binary mask. It's also worth exploring combining both outputs for tasks like semantic segmentation where detecting objects and understanding the scene at a pixel level is necessary.

If you're struggling with correct masks, you might experiment with various post-processing approaches on the segmentation output, like adjusting threshold values or applying morphological operations to refine the masks.

Here's a quick example on handling the segmentation output:

```python
import numpy as np

# Example to threshold mask output to binary image
mask_output = np.random.rand(1,160,160,32)  # Your Identity_1 output
average_mask = mask_output.mean(axis=-1).squeeze()  # Average across channels
threshold = 0.5  # Example threshold
binary_mask = (average_mask > threshold).astype(np.uint8)  # Thresholding

# Now, binary_mask is ready for visualization or combining with detection results
```

Hope this helps clarify how to use both types of outputs! Keep experimenting and happy coding! 🚀

**Comment by Layzerg** [comment]
Hi!
I have been working on training a segmentation model of single class objects in satellite images. The training has been progressing well so far. Recently, I was able to successfully train the model to segment two classes. In the future, I plan to expand the model's capabilities by training it to detect additional classes of objects. While there were no issues when it was a single class, there is now a need to divide the masks. Regrettably, the masks do not have a .cls attribute like the boxes do. Is there a way to obtain the class of each predicted mask from its corresponding bounding box while we iterating over the masks? Or there is some other way to obtain the class of segmentation mask?

**Comment by Habib-Rehmn** [comment]
> Hi! I have been working on training a segmentation model of single class objects in satellite images. The training has been progressing well so far. Recently, I was able to successfully train the model to segment two classes. In the future, I plan to expand the model's capabilities by training it to detect additional classes of objects. While there were no issues when it was a single class, there is now a need to divide the masks. Regrettably, the masks do not have a .cls attribute like the boxes do. Is there a way to obtain the class of each predicted mask from its corresponding bounding box while we iterating over the masks? Or there is some other way to obtain the class of segmentation mask?

Can you share more details on how you are doing this

**Comment by glenn-jocher** [comment]
@Habib-Rehmn hey there! 😊 Awesome to hear about your progress with segmentation on satellite images, especially moving into multi-class detection!

For mapping masks to their corresponding classes, if your segmentation masks are linked to detected bounding boxes (which contain class attributes), you can indeed match each mask to its class via their spatial overlap or index correlation.

If your setup produces separate lists/arrays for masks and detections (boxes with `.cls` attributes), maintaining order between these arrays allows you to associate each mask with a class directly through indexing.

An example approach, assuming `results.masks` and `results.boxes` are your outputs (with `results.boxes` containing the `.cls` attribute), could look like this:

```python
for i, mask in enumerate(results.masks):
    class_id = results.boxes[i].cls  # Assumes corresponding order
    print(f"Mask {i} class: {class_id}")
```

Alternatively, if masks are generated without a direct relation to detections, considering an association based on spatial overlap (intersecting a mask with bounding boxes and selecting the class of the highest overlapping box) might be necessary.

Remember, the exact solution could depend on how your model's output is structured. Sharing a bit more about your output format or the method you're using to generate masks and detection could help tailor the advice more closely to your needs! Keep up the great work! 🚀
